File: leetcode/0020-valid-parentheses/valid-parentheses.py
# Amir Mukhtarov, mukhtarov.amir.a@gmail.com
class Solution:
    def isValid(self, s: str) -> bool:
        h = {
            '}': '{',
            ']': '[',
            ')': '(',
        }

        stack = []

        for char in s:
            if char in h.values():
                stack.append(char)
            elif char in h:
                if not stack or stack.pop() != h[char]:
                    return False

        return len(stack) == 0


if __name__ == '__main__':
    s = "([])"
    solution = Solution()

    print(solution.isValid(s))

File: leetcode/0020-valid-parentheses/valid-parentheses.cpp
// Amir Mukhtarov, mukhtarov.amir.a@gmail.com
#include <iostream> 
#include <vector> 
#include <string> 
#include <algorithm> 
#include <unordered_map> 
#include <map> 
#include <unordered_set> 
#include <set> 
#include <queue> 
#include <stack> 
#include <deque> 
#include <cmath> 
#include <limits> 

using namespace std; 
 
class Solution {
public:
    bool isValid(string s) {
      stack<char> st;

      for (char c : s) {
        if (c == '{' || c == '[' || c == '(') {
          st.push(c);
        } else {
          if (st.empty()) return false;

          char top = st.top();
          st.pop();
          
          if (
            (c == ')' && top != '(') ||
            (c == '}' && top != '{') ||
            (c == ']' && top != '[')
          ) {
            return false;
          }

        }
      }
      return st.empty();
    }
};


int main(){ 
  string s = "([])";
  Solution solution;
  cout << solution.isValid(s) << endl;
  cout << "AAAAA" << endl;
  return 0;
}
